drivers/i2c/i2c_npcx_controller.c

```diff
@@ -0,0 +1,878 @@
+ *    +------+   +------------+ |  +------- ----+ |    +------- -------+  |
+ * +->| IDLE |-->| WAIT_START |--->| WRITE_FIFO |-+--->| WRITE_SUSPEND |--+
+ * |  +------+   +------------+    +------------+  Yes +---------------+  |
+ * |     Issue START     START completed                                  | No
+ * |      +---------------------------------------------------------------+
+ * |      |
+ * |      |                 All msg data received?  Is there next read msg?
+ * |      |                   +<-----------------+<-----------------------+
+ * |      |                   |       No         |                        | Yes
+ * |      |  +--------------+ |  +------- ---+   |    +------- ------+    |
+ * |      +--| WAIT_RESTART |--->| READ_FIFO |---+--->| READ_SUSPEND |----+
+ * |         +--------------+    +-----------+    Yes +--------------+    |
+ * |  Issue RESTART      RESTART completed                                | No
+ * |                                           +-----------+              |
+ * +-------------------------------------------| WAIT_STOP |<-------------+
+ *              STOP is completed              +-----------+  Issue STOP
+ *
+ */
+
+#include <assert.h>
+#include <drivers/i2c.h>
+#include <drivers/clock_control.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(i2c_npcx, LOG_LEVEL_ERR);
+
+/* I2C bus frequency */
+#define NPCX_I2C_BUS_SPEED_100KHZ 0
+#define NPCX_I2C_BUS_SPEED_400KHZ 1
+#define NPCX_I2C_BUS_SPEED_1MHZ   2
+
+/* I2C controller mode */
+#define NPCX_I2C_BANK_NORMAL 0
+#define NPCX_I2C_BANK_FIFO   1
+
+/* Timeout for device should be available after reset (SMBus spec. unit:ms) */
+#define I2C_MAX_TIMEOUT 35
+
+/* Timeout for SCL held to low by slave device . (SMBus spec. unit:ms). */
+#define I2C_MIN_TIMEOUT 25
+
+/* Default maximum time we allow for an I2C transfer (unit:ms) */
+#define I2C_TRANS_TIMEOUT K_MSEC(100)
+
+/*
+ * NPCX I2C module that supports FIFO mode has 32 bytes Tx FIFO and
+ * 32 bytes Rx FIFO.
+ */
+#define NPCX_I2C_FIFO_MAX_SIZE 32
+
+/* Valid bit fields in SMBST register */
+#define NPCX_VALID_SMBST_MASK ~(BIT(NPCX_SMBST_XMIT) | BIT(NPCX_SMBST_MASTER))
+
+/*
+ * Internal SMBus Interface driver states values, which reflect events
+ * which occurred on the bus
+ */
+enum smb_oper_state {
+	NPCX_I2C_IDLE,
+	NPCX_I2C_WAIT_START,
+	NPCX_I2C_WAIT_RESTART,
+	NPCX_I2C_WRITE_FIFO,
+	NPCX_I2C_WRITE_SUSPEND,
+	NPCX_I2C_READ_FIFO,
+	NPCX_I2C_READ_SUSPEND,
+	NPCX_I2C_WAIT_STOP,
+	NPCX_I2C_ERROR_RECOVERY,
+};
+
+/* I2C timing configuration for each i2c speed */
+struct npcx_i2c_timing_cfg {
+	uint8_t HLDT; /* i2c hold-time (Unit: clocks) */
+	uint8_t k1; /* k1 = SCL low-time (Unit: clocks) */
+	uint8_t k2; /* k2 = SCL high-time (Unit: clocks) */
+};
+
+/* Device config */
+struct i2c_ctrl_config {
+	uintptr_t base; /* i2c controller base address */
+	struct npcx_clk_cfg clk_cfg; /* clock configuration */
+	int irq; /* i2c controller irq */
```

Comment:
For MISRA these should be types with specific sizes, not unadorned `int`.  It'd also save space if they both fit as `uint8_t`.

Reply:
Done.